scripts/helper_scripts/unipept-database-rs/src/calculate_lcas/taxonomy.rs

```diff
@@ -0,0 +1,161 @@
+        }
+
+        let mut taxonomy = vec![Vec::new(); (max + 1) as usize];
+
+        for (key, value) in taxonomy_map {
+            taxonomy[key as usize] = value;
+        }
+
+        Ok(Taxonomy { taxonomy })
+    }
+
+    pub fn calculate_lcas(&self) -> Result<()> {
+        let reader = open_sin();
+
+        let mut current_sequence = String::new();
+        let mut taxa: Vec<i32> = Vec::new();
+
+        for (i, line) in reader.lines().enumerate() {
+            if i % 10000000 == 0 && i != 0 {
+                eprintln!("{}: {}", now_str(), i);
+            }
+
+            let line = line.context("error reading line from stdin")?;
+
+            let (sequence, taxon_id) =
+                line.split_once(SEPARATOR).context("error splitting line")?;
+            let taxon_id: i32 = taxon_id
+                .trim_end()
+                .parse()
+                .context("error parsing taxon id to int")?;
+
+            if current_sequence.is_empty() || current_sequence != sequence {
+                if !current_sequence.is_empty() {
+                    self.handle_lca(&current_sequence, self.calculate_lca(&taxa));
+                }
+
+                current_sequence = sequence.to_string();
+                taxa.clear();
+            }
+
+            taxa.push(taxon_id);
+        }
+
+        self.handle_lca(&current_sequence, self.calculate_lca(&taxa));
+        Ok(())
+    }
+
+    fn calculate_lca(&self, taxa: &[i32]) -> i32 {
+        let mut lca = 1;
+
+        let lineages: Vec<&Vec<i32>> = taxa
+            .iter()
+            .map(|x| &self.taxonomy[*x as usize])
+            .filter(|x| !x.is_empty())
+            .collect();
+
+        for rank in 0..RANKS {
+            let final_rank = rank;
+            let mut value = -1;
+
+            let iterator = lineages
+                .iter()
+                .map(|&x| x[final_rank as usize])
+                .filter(|&x| {
+                    if final_rank == GENUS || final_rank == SPECIES {
+                        x > 0
+                    } else {
+                        x >= 0
+                    }
+                });
+
+            let mut all_match = true;
+
+            // This was near-impossible to do with the iterators above,
+            // so we're using a simplified loop here
+            for item in iterator {
+                if value == -1 {
+                    value = item;
+                } else if item != value {
+                    all_match = false;
+                    break;
+                }
+            }
+
+            // If we found a new value that matched for all of them, use this as the new best
+            if value != -1 {
+                // If not everything matched, this is not a common ancestor anymore,
+                // so we can stop
+                if !all_match {
+                    break;
+                }
+
+                if value != 0 {
+                    lca = value;
+                }
+            }
```

Comment:
```suggestion
            for item in iterator {
                if value == -1 {
                    value = item;
                } else if item != value {
                    return lca;
                }
            }

            // If we found a new value that matched for all of them, use this as the new best
            if value > 0 {
                lca = value;
            }
```

I'm not 100% sure this is equivalent, but this is my thinking:
- Your line 137 breaks out of the outer loop, and the first statement after this loop is a return of `lca`, thus line 137 can be replaced by `return lca`
- Since the conditions are `value != -1` and `all_match == false`, the `return lca` can already occur at line `127`.
- I assumed `value` will never be less than `-1` if it has a value, so the conditionas at line 133 and 140 required to set `lca = value` can be replaced by a single condition `if value > 0`.

But I'm not familiar with the LCA algorithm, so maybe something is not correct.